src/ZEO/asyncio/_futures.pyx

```diff
@@ -0,0 +1,489 @@
+        return Future.result(self)
+
+
+IF UNAME_SYSNAME == "Windows":
+    cpdef switch_thread():
+        sleep(1e-6)
+ELSE:
+    cdef extern from "<sys/select.h>" nogil:
+        ctypedef struct fd_set
+        cdef struct timeval:
+           long tv_sec
+           unsigned long tv_usec
+        int select(int no, fd_set *rd, fd_set *wr, fd_set *ex,
+                   timeval *to) nogil
+    cpdef switch_thread():
+        cdef timeval timeout
+        timeout.tv_sec = 0; timeout.tv_usec = 1
+        with nogil:
+            select(0, NULL, NULL, NULL, &timeout)
+
+
+cdef class CoroutineExecutor:
+    """Execute a coroutine on behalf of a task.
+
+    No context support.
+    """
+    cdef object coro  # executed coroutine
+    cdef object task   # associated task
+    cdef object awaiting  # future we are waiting for
+    cdef bint   cancel_requested  # request to be canceled from .cancel
+    cdef object cancel_msg  # ^^^ cancellation message
+
+    def __init__(self, task, coro):
+        """execute *coro* on behalf of *task*."""
+        self.task = task  # likely introduces a reference cycle
+        self.coro = coro
+        self.awaiting = None
+        self.cancel_requested = False
+        self.cancel_msg = None
+
+    cpdef step(self):
+        await_result = None  # with what to wakeup suspended await
+        await_resexc = False # is it exception?
+        awaiting = self.awaiting
+        if awaiting is not None:
+            self.awaiting = None
+            assert awaiting.done()
+            try:
+                await_result = awaiting.result()
+            except BaseException as e:
+                await_result = e
+                await_resexc = True
+        if self.cancel_requested:
+            await_result = CancelledError()  if self.cancel_msg is None  else \
+                           CancelledError(self.cancel_msg)
+            await_resexc = True
+            self.cancel_requested = False
+            self.cancel_msg = None
+        try:
+            if not await_resexc:
+                result = self.coro.send(await_result)
+            else:
+                result = self.coro.throw(await_result)
+        except BaseException as e:
+            # we are done
+            task = self.task
+            self.task = None  # break reference cycle
+            if isinstance(e, (StopIteration, _GenReturn)):
+                if PY_MAJOR_VERSION < 3:
+                    v = getattr(e, 'value', None)  # e.g. no .value on plain return
+                    if hasattr(e, 'raised'):  # coroutines implemented inside trollius raise Return
+                        e.raised = True       # which checks it has been caught and complains if not
+                else:
+                    v = e.value
+                task.set_result(v)
+            elif isinstance(e, CancelledError):
+                if len(e.args) == 0:
+                    msg = getattr(awaiting, '_xasyncio_cancel_msg', None)  # see _cancel_future
```

Comment:
Thanks, @d-maurer. I originally wanted to avoid overlapping with names that original asyncio could use, and so used `_xasyncio` prefix for that (meaning that futures.py could be later renamed to xasyncio.py standing for "extra asyncio"). But you are right, if on `fut.cancel(msg)` we hit TypeError it seems safe to use the same attribute that modern asyncio.Future uses.

I corrected this in https://github.com/zopefoundation/ZEO/commit/0d0e16ab.